deployment/common/changeset/state/evm.go

```diff
@@ -82,16 +92,81 @@ func MaybeLoadMCMSWithTimelockState(env cldf.Environment, chainSelectors []uint6
 	return result, nil
 }
 
+// MergeAddressesFromBothSourcesEVMWithQualifier combines addresses from both DataStore and AddressBook making it backward compatible.
+// This version supports qualifiers for filtering DataStore addresses.
+// When a qualifier is specified, only DataStore addresses with that qualifier are returned (no AddressBook merge)
+// to ensure isolation between different deployments.
+func MergeAddressesFromBothSourcesEVMWithQualifier(env cldf.Environment, chainSelector uint64, qualifier string) (map[string]cldf.TypeAndVersion, error) {
+	// If a qualifier is specified, only use DataStore to ensure isolation between deployments
+	if qualifier != "" {
+		if env.DataStore != nil {
+			return LoadAddressesFromDataStore(env.DataStore, chainSelector, qualifier)
+		}
+		return nil, fmt.Errorf("DataStore not available but qualifier %s specified", qualifier)
+	}
+
+	// For backward compatibility without qualifier, merge both sources
+	// Start with addresses from AddressBook
+	addressBookAddresses := make(map[string]cldf.TypeAndVersion)
+	if addresses, err := env.ExistingAddresses.AddressesForChain(chainSelector); err == nil {
+		addressBookAddresses = addresses
+	} else if !errors.Is(err, cldf.ErrChainNotFound) {
+		return nil, fmt.Errorf("failed to load addresses from AddressBook: %w", err)
+	}
+
+	// Try to load addresses from DataStore (without qualifier for general case)
+	// Only try if DataStore is available
+	if env.DataStore != nil {
+		dataStoreAddresses, err := LoadAddressesFromDataStore(env.DataStore, chainSelector, "")
+		if err != nil {
+			// If DataStore has no addresses, just return AddressBook addresses
+			if strings.Contains(err.Error(), "no addresses found") {
+				return addressBookAddresses, nil
+			}
+			// Don't fail if DataStore is not working - fall back to AddressBook
+			return addressBookAddresses, nil
+		}
+
+		// Merge the two maps - DataStore addresses take precedence
+		mergedAddresses := make(map[string]cldf.TypeAndVersion)
+
+		// First add all AddressBook addresses
+		for addr, tv := range addressBookAddresses {
+			mergedAddresses[addr] = tv
+		}
+
+		// Then add DataStore addresses (overwriting any conflicts)
+		for addr, tv := range dataStoreAddresses {
+			mergedAddresses[addr] = tv
+		}
+
+		return mergedAddresses, nil
+	}
+
+	// If no DataStore, just return AddressBook addresses
+	return addressBookAddresses, nil
```

Comment:
/nit: I would have returned this earlier, with an early return:

```
if env.Datastore == nil {
	return addressBookAddresses, nil
}
```

then you can have the whole datastore dedicated section easier to read down at the end.